accepted/single-file/design.md

```diff
@@ -135,28 +135,17 @@ Once the single-file-publish tooling is added to the publish pipeline, other sta
 
 On Startup, the AppHost checks if it has embedded files. If so, it 
 
-* Processes configuration files such as `app.deps.json` , `app.runtimeconfig.json` from the embedded files, instead of looking for them on the disk next to the app.
-* Sets up the data-structures to locate pure managed assemblies within the bundle so that they can be loaded on-demand.
-* Extracts the other files (native libraries, ready-to-run compiled assemblies, and other data files to an *install-location*). Extraction of files is discussed in detail in this  [document](extract.md). 
-* Communicates the information about bundled and extracted files to the runtime, as explained in the section on dependency resolution.
+* Memory maps the entire bundle file.
+* Extracts the necessary files (ex: native binaries) to disk (if any) as explained in this [document](extract.md). 
+* Sets up data-structures so other components can access files embedded directly, as explained in this [document](direct.md)
 
 #### Dependency Resolution
 
 An app may choose to only embed some files (ex: due to licensing restrictions) and expect to pickup other dependencies from application-launch directory, nuget packages, etc. In order to resolve assemblies and native libraries, the embedded resources are probed first, followed by other probing paths. 
 
-The communication between the host and the runtime will continue to be through properties such as `TRUSTED_PLATFORM_ASSEMBLIES`, `NATIVE_DLL_SEARCH_DIRECTORIES`, etc.  These properties will be populated with entries from the bundled files (using a special marker) and the extracted files.
+#### App_Root Setting
 
-For example, if the TPA without single-exe publish is:
-`~/app/il.dll`; `~/app/r2r.dll`; `/usr/local/nuget/dep.dll`
-
-Where
-
-* `il.dll` is a pure managed assembly that is loaded directly from the bundle
-* `r2r.dll` is a ready-to-run assembly that is bundled and extract out to disk
-* `dep.dll` is an assembly that is not bundled with the app,
-
-the TPA with single-exe publish will be (assuming `?` is the bundle marker):
-`?/il.dll`; `/var/tmp/.net/app/e53xf3/r2r.dll`; `/usr/local/nuget/dep.dll`
+If files are extracted out to disk, the corehost `App_Root` is set to the directory where files are extracted. If all files in a bundle can be processed directly without need for extraction (an extraction directory is not created), and the `App_Root` is set to the directory where the AppHost resides.
```

Comment:
I know I may have suggested this during our discussions, but I am not sure whether it is a good idea after thinking about it some more. It creates non-desirable coupling where extracting even a single file will change the setting globally.

Reply:
The above formulation was based on the constraints that the application's files should be accessible from `AppContext.BaseDirectory`. However, when the all files are loaded from within the bundle, there is no extraction-directory for the `AppContext.BaseDirectory` to point at.

We have a few options::
* Always set `AppContext.BaseDirectory` to the directory containing the AppHost. Content extraction directory can be accessed (when applicable) via APIs such as those proposed here: https://github.com/dotnet/corefx/issues/37435 

* When running from bundle, set `AppContext.BaseDirectory`  to a distinct value, such as `null`, empty-string, or `:`.

* Always create an extraction directory (or have a default location when there are no files to be extracted) and point `AppContext.BaseDirectory`  to that directory.

* Set `AppContext.BaseDirectory` as described in this PR. But this has the shortcoming that the `AppContext.BaseDirectory` value changes based on the contents of the bundle -- which may be surprising for developers.

This problem is similar to the issue of setting `Assembly.Location` value.

Reply:
I've updated the PR to discuss the issue in terms of `AppContext.BaseDirectory`. Thanks.

Reply:
> Content extraction directory can be accessed (when applicable) via APIs such as those proposed 

I am worried about providing first class access to the extraction directory. Ideally, we would not have the extraction directory at all and we will hopefully get to a point in future where most apps do not need the extraction directory. Thus, the extraction directory should better be a hidden implementation detail. What about if we start with this: 

- Do not bundle application data files and leave them next to the .exe. (ie do what ASP.NET does today). Your app binary is a single file, but the app as a whole is not a single file.
- Or, attach data files to your app as managed resources. Access the data files via APIs for accessing managed resources.

Reply:
@jkotas, I've updated the PR, wrt `AppContext.Basedirectory` and `Assembly.Location` in this change https://github.com/dotnet/designs/pull/74/commits/9cde7edd390f74213e0d281ddc5d465107841ab0. Please take a look at it.

Reply:
Under these rules, certain apps that build artifacts (ex: asp.net using `deps.json` to read compiler configurations) will need to adapt to the situation with additional work -- ex: pack the `deps.json` file as a managed resource as well as bundled file, or device a different protocol to communicate the information without touching the `deps.json` file.